### PR TITLE
feat(telemetry): local JSONL sink with per-session files

### LINKED ISSUE #901: Telemetry: LocalJsonlSink with buffering, rotation, and cleanup
Part of the Telemetry Phase A rollout. See the RFC in Linear: [AIGOV-154](https://linear.app/codercom/issue/AIGOV-154/telemetry-audit-and-observability-instrumentation).

## Scope

* Implement `LocalJsonlSink` in `src/telemetry/sinks/localJsonlSink.ts` implementing the `TelemetrySink` interface from [AIGOV-243](https://linear.app/codercom/issue/AIGOV-243/telemetry-core-service-trace-and-servicecontainer-wiring).
* File layout under `globalStorageUri/telemetry/` with daily files `telemetry-YYYY-MM-DD.jsonl`.
* Buffer events in memory, flush every 15 seconds or when the buffer reaches 100 events. Both thresholds are configurable.
* Hard buffer cap of 500 events (configurable). When exceeded, drop oldest events and log a warning.
* Per-file size cap of 5 MB (configurable). When the current day's file exceeds the cap, create numbered segments (`telemetry-YYYY-MM-DD.1.jsonl`, `.2.jsonl`, and so on). Use `stat` before each flush, cached for the flush interval.
* Serialize events to snake_case JSON so they match the Phase B wire format without transformation.
* Use `fs.appendFile` so no file handles remain open between flushes.
* Write failures are logged via `Logger` but never thrown.
* `dispose()` flushes before returning.
* Cleanup on construction (runs at extension activation): delete files older than `maxAgeDays` (default 30), then remove oldest files if total size exceeds `maxSizeBytes` (default 100 MB). Same pattern as the SSH process log cleanup in `sshProcess.ts`.

## Tests

* Mock `fs/promises` with `memfs`, consistent with existing test patterns.
* Events are readable as valid JSONL after flush.
* Old files are removed on construction.
* Oversized storage is pruned oldest-first.
* Per-file cap creates a numbered segment when exceeded.
* Buffer overflow drops oldest events and logs a warning.
* `dispose()` persists all buffered events.

## Out of scope

* No instrumentation sites.
* No export command (separate ticket).

Depends on [AIGOV-243](https://linear.app/codercom/issue/AIGOV-243/telemetry-core-service-trace-and-servicecontainer-wiring).

### PR DESCRIPTION
**Pull request by EhabY**
## Summary

Adds `LocalJsonlSink`, the first concrete `TelemetrySink` on top of the Phase A foundation. Closes #901.

Each VS Code session writes its own files at:

```
globalStorage/telemetry/telemetry-YYYY-MM-DD-{sessionId8}.jsonl
                       telemetry-YYYY-MM-DD-{sessionId8}.1.jsonl   (size rotation)
                       telemetry-YYYY-MM-DD-{sessionId8}.2.jsonl
```

One writer per file means concurrent VS Code windows do not race on appends or rotation, so no lockfile is needed. Cleanup matches every session's files for shared retention.

The sink is best-effort: `write` is sync and never throws, all I/O happens in `flush`/`dispose`, errors are logged via the provided logger and never propagate to callers.

## Configuration

Tunables come from `coder.telemetry.localJsonl` (object-typed, not registered in `package.json` yet, matching the `coder.telemetry.level` convention). Defaults:

| Field | Default |
| --- | --- |
| `flushIntervalMs` | 15000 |
| `flushBatchSize` | 100 |
| `bufferLimit` | 500 |
| `maxFileBytes` | 5242880 |
| `maxAgeDays` | 30 |
| `maxTotalBytes` | 104857600 |

The sink subscribes via `watchConfigurationChanges`, so changes apply without a window reload. Invalid or missing fields fall back to defaults.

## Why per-session naming over a shared file

`fs.appendFile` is only atomic per `write()` syscall up to PIPE_BUF (~4KB). A 100-event batch can split across syscalls, letting two windows interleave bytes mid-line. `proper-lockfile` exists in the repo (used by `BinaryLock`) but its workload is one-shot and waits-or-skips; a hot-path lock with stale-timeout tradeoffs is a worse fit than just having one writer per file.

### CODE REVIEW

File: src/settings/telemetry.ts [hunk 1]
```diff
@@ -0,0 +1,69 @@
+import type { WorkspaceConfiguration } from "vscode";
```
[the remark this answers is not in the document]
Comment (EhabY) [reply src/settings/telemetry.ts:1]: Added `test/unit/settings/telemetry.test.ts`. Covers off/local/invalid/non-string for the level, and non-object raw + per-field invalid types (zero, negative, NaN, string, boolean) for the JSONL config.

File: test/unit/telemetry/sinks/localJsonlSink.test.ts [hunk 2]
```diff
@@ -0,0 +1,357 @@
...
+import type * as fs from "node:fs";
+
+import type { TelemetryEvent } from "@/telemetry/event";
+
+vi.mock("node:fs/promises", async () => {
+	const memfs: { fs: typeof fs } = await vi.importActual("memfs");
+	return memfs.fs.promises;
+});
+
+const SETTING_NAME = "coder.telemetry.localJsonl";
```
[the remark this answers is not in the document]
Comment (EhabY) [reply test/unit/telemetry/sinks/localJsonlSink.test.ts]: Done, dropped `SETTING_NAME` for the imported `LOCAL_JSONL_SETTING`.

File: src/telemetry/sinks/localJsonlSink.ts [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+				const next = changes.get(LOCAL_JSONL_SETTING) as
+					| LocalJsonlConfig
+					| undefined;
+				if (next) {
+					sink.#config = next;
+				}
+			},
+		);
+		sink.#scheduleNextFlush();
+		void sink.#cleanupOldFiles();
```
[the remark this answers is not in the document]
Comment (EhabY) [reply src/telemetry/sinks/localJsonlSink.ts:81]: Fixed. `#append` now calls `fs.mkdir({recursive: true})` before `appendFile`. Idempotent and cheap, no flag needed. Cleanup no longer mkdirs since `cleanupFiles` returns early on ENOENT readdir.

File: src/telemetry/sinks/localJsonlSink.ts [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+			this.#baseDir,
+			`${FILE_PREFIX}${next.date}-${this.#sessionSlug}${seg}${FILE_SUFFIX}`,
+		);
+		try {
+			await fs.appendFile(target, payload, "utf8");
+			this.#current = { ...next, size: next.size + payload.length };
+			this.#overflowWarned = false;
+		} catch (err) {
+			// Leave #current and #overflowWarned alone: the next flush re-evaluates
+			// rotation, and the warn flag fires again on the next overflow burst.
```
[the remark this answers is not in the document]
Comment (EhabY) [reply src/telemetry/sinks/localJsonlSink.ts]: Fixed. The catch resets `#overflowWarned` too, so the next overflow burst still warns when flushes keep failing.

File: test/unit/telemetry/sinks/localJsonlSink.test.ts [hunk 5]
```diff
@@ -0,0 +1,357 @@
...
+		sink.write(makeEvent());
+		sink.write(makeEvent());
+		expect(vol.existsSync(todaysFile())).toBe(false);
+
+		sink.write(makeEvent());
+		await vi.waitFor(() => expect(vol.existsSync(todaysFile())).toBe(true));
+		expect(readJsonl(todaysFile())).toHaveLength(3);
+	});
+
+	it("drops the oldest events when the buffer exceeds bufferLimit", async () => {
```
[the remark this answers is not in the document]
Comment (EhabY) [reply test/unit/telemetry/sinks/localJsonlSink.test.ts:129]: Combined into one test that runs three overflow bursts with the second flush mocked to fail and asserts three warnings.

File: src/util/fileCleanup.ts [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+			}),
+		);
+
+		const successful = deleted.filter((name) => name !== null);
+		if (successful.length > 0) {
+			logger.debug(
+				`Cleaned up ${successful.length} ${fileType}(s): ${successful.join(", ")}`,
+			);
+		}
+	} catch {
```
[the remark this answers is not in the document]
Comment (EhabY) [reply src/util/fileCleanup.ts]: Fixed per your suggestion. Outer try is around `readdir` only; non-ENOENT errors log at debug. Added a test that mocks `readdir` to reject with EACCES and asserts `pick` is not called.

File: src/core/container.ts [hunk 7]
```diff
@@ -90,7 +91,18 @@ export class ServiceContainer implements vscode.Disposable {
 			context.extensionUri,
 			this.logger,
 		);
-		this.telemetryService = new TelemetryService(context, [], this.logger);
+		const localJsonlSink = LocalJsonlSink.start(
+			{
+				baseDir: this.pathResolver.getTelemetryPath(),
+				sessionId: vscode.env.sessionId,
```
[the remark this answers is not in the document]
Comment (EhabY) [reply src/core/container.ts:97]: Registered `coder.telemetry.level` in `package.json` with description, enum descriptions, default `local`, and the `telemetry` tag, so it shows up in the settings UI. Default stays `local` since events are local-only and never leave the machine.

File: src/telemetry/sinks/localJsonlSink.ts [hunk 8]
```diff
@@ -0,0 +1,293 @@
...
+			this.#logger.warn(`Telemetry sink '${this.name}' flush failed`, err);
```
[the remark this answers is not in the document]
Comment (EhabY) [reply src/telemetry/sinks/localJsonlSink.ts]: Done. The catch now logs `flush failed, ${lines.length} event(s) discarded`.

File: src/telemetry/sinks/localJsonlSink.ts [hunk 9]
```diff
@@ -0,0 +1,293 @@
...
+		const payload = lines.join("");
+		const next = this.#nextFile(payload.length);
+		const seg = next.segment > 0 ? `.${next.segment}` : "";
+		const target = path.join(
+			this.#baseDir,
+			`${FILE_PREFIX}${next.date}-${this.#sessionSlug}${seg}${FILE_SUFFIX}`,
+		);
+		try {
+			await fs.appendFile(target, payload, "utf8");
+			this.#current = { ...next, size: next.size + payload.length };
```
[the remark this answers is not in the document]
Comment (EhabY) [reply src/telemetry/sinks/localJsonlSink.ts]: Fixed. `payload.length` is now `Buffer.byteLength(payload, "utf8")` at both the rotation check and the size update.

File: src/telemetry/sinks/localJsonlSink.ts [hunk 10]
```diff
@@ -0,0 +1,293 @@
...
+	readonly #baseDir: string;
+	readonly #sessionSlug: string;
+	readonly #logger: Logger;
+	readonly #buffer: string[] = [];
+	#config: LocalJsonlConfig;
+	#configWatcher: vscode.Disposable | null = null;
+	#flushTimer: NodeJS.Timeout | null = null;
+	#flushChain: Promise<void> = Promise.resolve();
+	#hasQueued = false;
+	#current: CurrentFile = { date: "", segment: 0, size: 0 };
```
[the remark this answers is not in the document]
Comment (EhabY) [reply src/telemetry/sinks/localJsonlSink.ts:56]: Fixed. On the first flush after activation, `#nextFile` stats segment 0 and seeds `size` from it. ENOENT falls through to zero. Higher pre-existing segments are a corner case (millions of events in one day) so I followed the simpler suggestion you posted.

File: test/unit/telemetry/sinks/localJsonlSink.test.ts [hunk 11]
```diff
@@ -0,0 +1,357 @@
...
+		setup({ maxAgeDays: 30 });
+
+		await vi.waitFor(() =>
+			expect(vol.readdirSync(BASE_DIR)).toEqual([
+				`telemetry-${today}-bbbb2222.jsonl`,
+			]),
+		);
+	});
+
+	it("trims oldest files when total size exceeds maxTotalBytes", async () => {
```
[the remark this answers is not in the document]
Comment (EhabY) [reply test/unit/telemetry/sinks/localJsonlSink.test.ts:267]: Added a second case with `maxTotalBytes: 2500` against three 2000-byte files. The loop has to delete two of them, so a hardcoded break after one would fail.

File: test/unit/telemetry/sinks/localJsonlSink.test.ts [hunk 12]
```diff
@@ -0,0 +1,357 @@
...
+			sink.write(makeEvent());
+		}
+		await sink.flush();
+
+		expect(readJsonl(todaysFile()).map((l) => l.event_sequence)).toEqual([
+			3, 4, 5, 6, 7,
+		]);
+	});
+
+	it("flushes pending events on dispose", async () => {
```
[the remark this answers is not in the document]
Comment (EhabY) [reply test/unit/telemetry/sinks/localJsonlSink.test.ts:182]: Added an `ignores writes after dispose` test that writes events post-dispose and asserts the file still only contains the pre-dispose ones.

File: src/util/fileCleanup.ts [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+	size: number;
+}
+
+export interface FileCleanupOptions {
+	/** Noun used in log messages, e.g. "telemetry file". */
+	fileType: string;
+	/** Name-based filter applied before stat to skip unrelated entries. */
+	match?: (name: string) => boolean;
+	/** Picks files to delete from the stat'd survivors of `match`. */
+	pick: (files: FileCleanupCandidate[], now: number) => Array<{ name: string }>;
```
[the remark this answers is not in the document]
Comment (EhabY) [reply src/util/fileCleanup.ts]: Fixed. `cleanupFiles` runs `file.name` through `path.basename` before joining. Added a test with a sibling file outside the cleanup directory and asserts it survives a pick that returns `../whatever`.

File: src/telemetry/sinks/localJsonlSink.ts [hunk 14]
```diff
@@ -0,0 +1,293 @@
...
+		let line: string;
+		try {
+			line = serializeEvent(event);
+		} catch (err) {
+			this.#logger.warn(`Telemetry sink '${this.name}' serialize failed`, err);
+			return;
+		}
+		this.#buffer.push(line);
+
+		if (this.#buffer.length > this.#config.bufferLimit) {
```
[the remark this answers is not in the document]
Comment (EhabY) [reply src/telemetry/sinks/localJsonlSink.ts]: Went with a warn rather than a clamp. `start()` and the watcher both call `warnIfBufferTooSmall`, which logs when `bufferLimit < flushBatchSize` and explains how to fix it. Clamping silently rewrites the user's value, which felt worse than surfacing the misconfig.

File: src/settings/telemetry.ts [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+
+import type { TelemetryLevel } from "../telemetry/event";
+
+export const TELEMETRY_LEVEL_SETTING = "coder.telemetry.level";
+export const LOCAL_JSONL_SETTING = "coder.telemetry.localJsonl";
```
[the remark this answers is not in the document]
Comment (EhabY) [reply src/settings/telemetry.ts]: Done in this PR. Both keys are now in `contributes.configuration.properties`.

File: src/telemetry/service.ts [hunk 16]
```diff
@@ -56,11 +61,13 @@ export class TelemetryService implements vscode.Disposable {
 		this.#configWatcher = watchConfigurationChanges(
 			[{ setting: TELEMETRY_LEVEL_SETTING, getValue: readLevel }],
 			(changes) => {
-				const raw = changes.get(TELEMETRY_LEVEL_SETTING);
-				if (!isTelemetryLevel(raw)) {
+				const next = changes.get(TELEMETRY_LEVEL_SETTING) as
```
[the remark this answers is not in the document]
Comment (EhabY) [reply src/telemetry/service.ts:64]: The cast just narrows the `Map<string, unknown>` value type. `getValue` is typed to return `TelemetryLevel` / `LocalJsonlConfig` and the read functions already validate, so by construction nothing else can land in that map. The runtime guard would only re-check what the read site already enforces.

File: src/telemetry/sinks/localJsonlSink.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (EhabY) [reply src/telemetry/sinks/localJsonlSink.ts:81]: Intentional. `dispose` drains writes; cleanup is best-effort and one-shot per activation. In-flight unlinks just retry on next startup, and disposed `LogOutputChannel` writes are no-ops in VS Code as you noted.

File: src/telemetry/sinks/localJsonlSink.ts [hunk 17]
```diff
@@ -0,0 +1,293 @@
...
+		if (
+			this.#current.size > 0 &&
+			this.#current.size + payloadSize > this.#config.maxFileBytes
+		) {
+			return { date: today, segment: this.#current.segment + 1, size: 0 };
+		}
+		return this.#current;
+	}
+
+	async #cleanupOldFiles(): Promise<void> {
```
[the remark this answers is not in the document]
Comment (EhabY) [reply src/telemetry/sinks/localJsonlSink.ts:244]: Right, this matches typical log-rotation (rotate at startup, not on write). `maxTotalBytes` is a startup ceiling, not a continuous cap. We can move it behind the flush timer if real users blow past it, but per-flush cleanup has a real cost.

File: src/telemetry/sinks/localJsonlSink.ts [hunk 18]
```diff
@@ -0,0 +1,336 @@
...
+
+	async #append(payload: string, payloadBytes: number): Promise<void> {
+		// mkdir is idempotent with recursive:true; cheap enough to do per flush.
+		await fs.mkdir(this.#baseDir, { recursive: true });
+		const next = await this.#nextFile(payloadBytes);
+		await fs.appendFile(this.#segmentPath(next), payload, "utf8");
+		this.#current = { ...next, size: next.size + payloadBytes };
+	}
+
+	async #nextFile(payloadSize: number): Promise<CurrentFile> {
```
[the remark this answers is not in the document]
Comment (EhabY) [reply src/telemetry/sinks/localJsonlSink.ts:216]: Fixed. `#nextFile` now applies the rotation check after `#seedFromDisk` returns, so a full segment 0 from a prior EH activation rotates to segment 1 instead of overrunning.

File: test/unit/telemetry/sinks/localJsonlSink.test.ts [hunk 19]
```diff
@@ -0,0 +1,424 @@
...
+			a.sink.write(a.makeEvent());
+			b.sink.write(b.makeEvent());
+		}
+		await Promise.all([a.sink.flush(), b.sink.flush()]);
+
+		expect(readJsonl(todaysFile(SESSION_SLUG))).toHaveLength(5);
+		expect(readJsonl(todaysFile("ffeeddcc"))).toHaveLength(5);
+	});
+
+	it("coalesces concurrent flush requests into at most two appendFile calls", async () => {
```
[the remark this answers is not in the document]
Comment (EhabY) [reply test/unit/telemetry/sinks/localJsonlSink.test.ts]: Added `expect(p3).toBe(p2)` to the coalescing test. The third caller while a flush is queued must share the same promise, which is the observable proof of the guard.

File: src/settings/telemetry.ts [hunk 20]
```diff
@@ -0,0 +1,69 @@
...
+		maxAgeDays: positiveNumber(obj.maxAgeDays, LOCAL_JSONL_DEFAULTS.maxAgeDays),
+		maxTotalBytes: positiveNumber(
+			obj.maxTotalBytes,
+			LOCAL_JSONL_DEFAULTS.maxTotalBytes,
+		),
+	};
+}
+
+function positiveNumber(value: unknown, fallback: number): number {
+	return typeof value === "number" && value > 0 ? value : fallback;
```
[the remark this answers is not in the document]
Comment (EhabY) [reply src/settings/telemetry.ts]: Fixed. `readLocalJsonlConfig` now enforces per-field minimums via `numberAtLeast`. Below-floor values fall back to the default. Also tightened floors that were meaninglessly low: bufferLimit 1 to 10, maxFileBytes 1024 to 4096, maxTotalBytes 1024 to 4096.

File: test/unit/telemetry/sinks/localJsonlSink.test.ts (hunk 19 above)
[the remark this answers is not in the document]
Comment (EhabY) [reply test/unit/telemetry/sinks/localJsonlSink.test.ts]: Added `keeps a single oversized payload in segment 0 instead of rotating`. Writes one event larger than `maxFileBytes` and asserts segment 1 is never created.

File: src/telemetry/sinks/localJsonlSink.ts [hunk 21]
```diff
@@ -0,0 +1,336 @@
...
+async function statBytes(target: string): Promise<number> {
+	try {
+		return (await fs.stat(target)).size;
+	} catch {
+		return 0;
+	}
+}
+
+function todayUtc(): string {
+	return new Date().toISOString().slice(0, 10);
```
[the remark this answers is not in the document]
Comment (EhabY) [reply src/telemetry/sinks/localJsonlSink.ts:301]: Added `warns when bufferLimit is below flushBatchSize` that asserts the warning fires at start.

File: src/telemetry/sinks/localJsonlSink.ts [hunk 22]
```diff
@@ -0,0 +1,336 @@
...
+			}
+			toDelete.push({ name: file.name });
+			totalBytes -= file.size;
+		}
+		return toDelete;
+	};
+}
+
+async function statBytes(target: string): Promise<number> {
+	try {
```
[the remark this answers is not in the document]
Comment (EhabY) [reply src/telemetry/sinks/localJsonlSink.ts:290]: Fixed. `statBytes` now logs non-ENOENT errors at debug, matching the pattern in `cleanupFiles`. Returns 0 either way so the caller can proceed.

File: src/telemetry/sinks/localJsonlSink.ts [hunk 23]
```diff
@@ -0,0 +1,336 @@
...
+	public static start(
+		opts: LocalJsonlSinkOptions,
+		logger: Logger,
+	): LocalJsonlSink {
+		const config = readLocalJsonlConfig(vscode.workspace.getConfiguration());
+		warnIfBufferTooSmall(config, logger);
+		const sink = new LocalJsonlSink(opts, logger, config);
+		sink.#configWatcher = sink.#watchConfig();
+		sink.#scheduleNextFlush();
+		void sink.#cleanupOldFiles();
```
[the remark this answers is not in the document]
Comment (EhabY) [reply src/telemetry/sinks/localJsonlSink.ts:81]: Fixed. The cleanup `match` now excludes the current session's files via the session slug. Cleanup operates only on other sessions' files, so it cannot race with our own appends.

File: src/telemetry/sinks/localJsonlSink.ts [hunk 24]
```diff
@@ -0,0 +1,336 @@
...
+	}
+
+	// Chained on completion (not setInterval) so flushes never overlap or pile up.
+	#scheduleNextFlush(): void {
+		if (this.#disposed) {
+			return;
+		}
+		this.#flushTimer = setTimeout(() => {
+			this.flush()
+				.catch((err) => {
```
[the remark this answers is not in the document]
Comment (EhabY) [reply src/telemetry/sinks/localJsonlSink.ts]: Removed. `#doFlush` catches every I/O error and `flushChain.then(run, run)` never propagates a rejection, so the .catch was unreachable.

File: src/telemetry/sinks/localJsonlSink.ts [hunk 25]
```diff
@@ -0,0 +1,336 @@
...
+					this.#logger.warn(
+						`Telemetry sink '${this.name}' scheduled flush failed`,
+						err,
+					);
+				})
+				.finally(() => this.#scheduleNextFlush());
+		}, this.#config.flushIntervalMs);
```
[the remark this answers is not in the document]
Comment (EhabY) [reply src/telemetry/sinks/localJsonlSink.ts:185]: Fixed. The watcher cancels the running timer and reschedules when `flushIntervalMs` changes, so an interval reduction takes effect immediately. Matches what OpenTelemetry, Datadog, and Sentry do (reschedule, no immediate flush).

**Review by jakehwll — APPROVED** [review]
LGTM 🙂
